AAInstrumentation/maximal_relaxation.py

```diff
@@ -0,0 +1,306 @@
+                upper_bound = new_count
+                break
+
+            if new_count == upper_bound:
+                # If there are equally many queries, we are done
+                break
+
+            print("New prefix: " + str(prefix))
+
+            # If there are more queries, we try to relax the new ones
+            lower_bound = len(prefix)
+            upper_bound = new_count
+
+        print("Final list: " + str(prefix))
+        print("with size: " + str(len(prefix)) + " of " + str(upper_bound))
+
+
+class SequentialMaximalRelaxationDriver(MaximalRelaxationDriver):
+    def relax(
+        self,
+        file_name: Path,
+        prefix,
+        total_lower_bound,
+        total_upper_bound,
+    ):
+        """
+        This function works batch-wise. It starts with a batch with size
+        proc_count and checks if all of the queries can be relaxed. If so, it
+        repeats with the next batch. If query i could not be relaxed, it leaves
+        query i precise and repeats with a batch starting at i+1. If multiple
+        queries could not be relaxed, it treats the smallest index as i in the
+        single query case.
+        """
+
+        os.makedirs(str((self.instr_dir / file_name).parent), exist_ok=True)
+
+        self.original_hash = self.get_groundtruth_hash(file_name)
+
+        lower_bound = 0
+
+        while True:
+            upper_bound = min(total_upper_bound, lower_bound + self.proc_count)
+            if lower_bound == upper_bound:
+                break
+
+            with Pool(self.proc_count) as p:
+                p.starmap(
+                    self.relax_one_index,
+                    [
+                        (
+                            prefix,
+                            file_name,
+                            lower_bound,
+                            i + 1,
+                        )
+                        for i in range(lower_bound, upper_bound)
+                    ],
+                )
+
+            for i in range(lower_bound, upper_bound):
+                curr_hash = self.get_hash(
+                    self.instr_dir.joinpath(
+                        file_name.parent, str(i + 1) + str(file_name.stem) + ".o"
+                    )
+                )
+                if curr_hash != self.original_hash:
+                    lower_bound = i + 1
+                    break
+
+                prefix.append(i)
+                lower_bound = i + 1
+
+        return prefix
+
+
+class LogMaximalRelaxationDriver(MaximalRelaxationDriver):
+    def relax(
+        self,
+        file_name: Path,
+        prefix,
+        lower_bound,
+        upper_bound,
+    ):
+        return self.rec_relax(file_name, prefix, lower_bound, upper_bound)
+
+    def rec_relax(
+        self,
+        file_name,
+        prefix,
+        lower_bound,
+        upper_bound,
+    ) -> list[int]:
+        """Recursivle relaxes the queries from [lower_bound, upper_bound), with
+        the prefix."""
```

Comment:
would it make sense to `if prefix; assert prefix[-1] < lower_bound`?

Reply:
Shua